src/cryptonote_core/service_node_swarm.cpp

```diff
@@ -0,0 +1,272 @@
+    if (it != swarm_to_snodes.end()) {
+      unassigned_snodes = it->second;
+      swarm_to_snodes.erase(it);
+    }
+
+    LOG_PRINT_L3("calc_swarm_changes. swarms: " << swarm_to_snodes.size() << ", regs: " << unassigned_snodes.size());
+
+    /// 0. Ensure there is always 1 swarm
+    if (swarm_to_snodes.size() == 0)
+    {
+      const auto new_swarm_id = get_new_swarm_id(mersenne_twister, {});
+      swarm_to_snodes.insert({new_swarm_id, {}});
+      LOG_PRINT_L2("Created initial swarm " << new_swarm_id);
+    }
+
+    /// 1. Assign new registered snodes
+    assign_snodes(unassigned_snodes, swarm_to_snodes, mersenne_twister);
+    LOG_PRINT_L2("After assignment:");
+    for (const auto &entry : swarm_to_snodes)
+    {
+      LOG_PRINT_L2(entry.first << ": " << entry.second.size());
+    }
+
+    /// 2. *Robin Hood Round* steal snodes from wealthy swarms and give them to the poor
+    {
+      std::vector<swarm_size> sorted_swarm_sizes;
+      calc_swarm_sizes(swarm_to_snodes, sorted_swarm_sizes);
+      bool insufficient_excess = false;
+      for (const auto& swarm : sorted_swarm_sizes)
+      {
+        /// we have processed all the starving swarms
+        if (swarm.size >= MIN_SWARM_SIZE)
+          break;
+
+        auto& poor_swarm_snodes = swarm_to_snodes.at(swarm.swarm_id);
+        do
+        {
+          const size_t percentile_index = STEALING_SWARM_UPPER_PERCENTILE * (sorted_swarm_sizes.size() - 1) / 100;
+          /// -1 since we will only consider swarm sizes strictly above percentile_value
+          size_t percentile_value = sorted_swarm_sizes.at(percentile_index).size - 1;
+          percentile_value = std::max(MIN_SWARM_SIZE, percentile_value);
+          size_t excess;
+          std::vector<excess_pool_snode> excess_pool;
+          get_excess_pool(percentile_value, swarm_to_snodes, excess_pool, excess);
+          /// If we can't save the swarm, don't bother continuing
+          const size_t deficit = MIN_SWARM_SIZE - poor_swarm_snodes.size();
+          insufficient_excess = (excess < deficit);
+          if (insufficient_excess)
+            break;
+          const auto& excess_snode = pick_from_excess_pool(excess_pool, mersenne_twister);
+          remove_excess_snode_from_swarm(excess_snode, swarm_to_snodes);
+          /// Add public key to poor swarm
+          poor_swarm_snodes.push_back(excess_snode.public_key);
+          LOG_PRINT_L2("Stolen 1 snode from " << excess_snode.public_key << " and donated to " << swarm.swarm_id);
+        } while (poor_swarm_snodes.size() < MIN_SWARM_SIZE);
+
+        /// If there is not enough excess for the current swarm,
+        /// there isn't either for the next one since the swarms are sorted
+        if (insufficient_excess)
+          break;
+      }
+    }
+
+    /// 3. New swarm creation
+    create_new_swarm_from_excess(swarm_to_snodes, mersenne_twister);
+
+    /// 4. If there is a swarm with less than MIN_SWARM_SIZE, decommission that swarm.
```

Comment:
Perhaps we should decommission swarms before we try to create a new swarm, so that we can create use them to create a new swarm. Actually now I'm thinking we could even use them to fill other would-be decommissioned swarms and thus prevent them from becoming decommissioned.